File: packages/js-sdk/src/types/bundle.ts
/**
 * TypeScript types for TurboQuote — Bundle entity and request/response types
 */

import type { BillingFrequency, Currency, DiscountType, PaginationParams, PaginatedResponse } from './quote-shared';
import type { Product } from './product';

// ============================================
// DOMAIN TYPES
// ============================================

export type BundleItemStatus = 'active' | 'product_deleted' | 'product_unavailable' | 'currency_mismatch';

export interface BundleItem {
  id: string;
  orgId: string;
  bundleId: string;
  productId: string;
  quantity: number;
  unitPrice: number;
  discountPercent: number;
  discountType: DiscountType | null;
  discountAmount: number | null;
  finalPrice: number;
  cost: number | null;
  billingFrequency: BillingFrequency;
  itemStatus: BundleItemStatus;
  isActive: boolean;
  createdBy: string | null;
  createdOn: string;
  updatedOn: string;
  product?: Product;
}

export interface Bundle {
  id: string;
  orgId: string;
  name: string;
  description: string | null;
  sku: string | null;
  categoryId: string | null;
  bundleDiscountPercent: number;
  bundleDiscountType: DiscountType | null;
  bundleDiscountAmount: number | null;
  totalListPrice: number;
  totalFinalPrice: number;
  totalCost: number;
  currency: Currency;
  showItemsToEndUser: boolean;
  showInCatalog: boolean;
  syncWithProducts: boolean;
  isActive: boolean;
  createdBy: string | null;
  createdOn: string;
  updatedOn: string;
  items?: BundleItem[];
  category?: { id: string; name: string; categoryType: string };
}

// ============================================
// REQUEST TYPES
// ============================================

export interface BundleItemInput {
  productId: string;
  unitPrice: number;
  billingFrequency: BillingFrequency;
  quantity?: number;
  discountPercent?: number;
  discountType?: DiscountType;
  discountAmount?: number;
  finalPrice?: number;
  cost?: number | null;
}

export interface CreateBundleRequest {
  name: string;
  categoryId: string;
  items?: BundleItemInput[];
  description?: string | null;
  sku?: string | null;
  bundleDiscountPercent?: number;
  bundleDiscountType?: DiscountType;
  bundleDiscountAmount?: number;
  currency?: Currency;
  showItemsToEndUser?: boolean;
  showInCatalog?: boolean;
  syncWithProducts?: boolean;
}

export interface UpdateBundleRequest {
  name?: string;
  items?: BundleItemInput[];
  description?: string | null;
  sku?: string | null;
  categoryId?: string;
  bundleDiscountPercent?: number;
  bundleDiscountType?: DiscountType;
  bundleDiscountAmount?: number;
  currency?: Currency;
  showItemsToEndUser?: boolean;
  showInCatalog?: boolean;
  syncWithProducts?: boolean;
}

export interface ListBundlesOptions extends PaginationParams {
  categoryIds?: string | string[];
  currency?: Currency;
  showInCatalog?: boolean;
}

// ============================================
// RESPONSE TYPES
// ============================================

export interface BundleListResponse extends PaginatedResponse<Bundle> {
  totalBundles: number;
  activeBundles: number;
  totalCategories: number;
  catalogValue: number;
}

File: packages/py-sdk/src/turbodocx_sdk/types/bundle.py
"""
TypedDict definitions for TurboQuote -- Bundle entity.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional
from typing_extensions import Literal, TypedDict

from .product import Product
from .quote_shared import DiscountType

BundleItemStatus = Literal["active", "product_deleted", "product_unavailable", "currency_mismatch"]


class BundleItem(TypedDict, total=False):
    id: str
    orgId: str
    bundleId: str
    productId: str
    quantity: int
    unitPrice: float
    discountPercent: float
    finalPrice: float
    cost: Optional[float]
    billingFrequency: str  # BillingFrequency literal
    itemStatus: str  # BundleItemStatus literal
    isActive: bool
    createdBy: Optional[str]
    createdOn: str
    updatedOn: str
    product: Product


class Bundle(TypedDict, total=False):
    id: str
    orgId: str
    name: str
    description: Optional[str]
    sku: Optional[str]
    categoryId: Optional[str]
    bundleDiscountPercent: float
    totalListPrice: float
    totalFinalPrice: float
    totalCost: float
    currency: str  # Currency literal
    showItemsToEndUser: bool
    showInCatalog: bool
    syncWithProducts: bool
    isActive: bool
    createdBy: Optional[str]
    createdOn: str
    updatedOn: str
    items: List[BundleItem]
    category: Dict[str, Any]


class _BundleItemInputRequired(TypedDict):
    productId: str
    unitPrice: float
    billingFrequency: str  # BillingFrequency literal


class BundleItemInput(_BundleItemInputRequired, total=False):
    quantity: int
    discountType: DiscountType
    discountPercent: float
    discountAmount: float
    finalPrice: float
    cost: Optional[float]


class _CreateBundleRequestRequired(TypedDict):
    name: str
    categoryId: str


class CreateBundleRequest(_CreateBundleRequestRequired, total=False):
    items: List[BundleItemInput]
    description: Optional[str]
    sku: Optional[str]
    bundleDiscountType: DiscountType
    bundleDiscountPercent: float
    bundleDiscountAmount: float
    currency: str  # Currency literal
    showItemsToEndUser: bool
    showInCatalog: bool
    syncWithProducts: bool


class UpdateBundleRequest(TypedDict, total=False):
    name: str
    items: List[BundleItemInput]
    description: Optional[str]
    sku: Optional[str]
    categoryId: str
    bundleDiscountType: DiscountType
    bundleDiscountPercent: float
    bundleDiscountAmount: float
    currency: str  # Currency literal
    showItemsToEndUser: bool
    showInCatalog: bool
    syncWithProducts: bool


class ListBundlesOptions(TypedDict, total=False):
    limit: int
    offset: int
    query: str
    categoryIds: Any  # string | string[]
    currency: str  # Currency literal
    showInCatalog: bool


class BundleListResponse(TypedDict, total=False):
    results: List[Bundle]
    totalRecords: int
    totalBundles: int
    activeBundles: int
    totalCategories: int
    catalogValue: float
